tracing-subscriber/src/filter/directive.rs

```diff
@@ -284,23 +286,53 @@ impl Default for Directive {
 
 impl PartialOrd for Directive {
     fn partial_cmp(&self, other: &Directive) -> Option<Ordering> {
-        match (self.has_name(), other.has_name()) {
-            (true, false) => return Some(Ordering::Greater),
-            (false, true) => return Some(Ordering::Less),
-            _ => {}
-        }
-
-        match (self.fields.len(), other.fields.len()) {
-            (a, b) if a == b => {}
-            (a, b) => return Some(a.cmp(&b)),
+        // We attempt to order directives by how "specific" they are. This
+        // ensures that we try the most specific directives first when
+        // attempting to match a piece of metadata.
+
+        // First, we compare based on whether a target is specified, and the
+        // lengths of those targets if both have targets.
+        let target_lengths = match (self.target.as_ref(), other.target.as_ref()) {
+            (Some(_), None) => Ordering::Greater,
+            (None, Some(_)) => Ordering::Less,
+            (Some(a), Some(b)) => a.len().cmp(&b.len()),
+            (None, None) => Ordering::Equal,
+        };
+        // Next compare based on the presence of span names.
+        let has_span_name = || match (self.in_span.is_some(), other.in_span.is_some()) {
+            (true, false) => Ordering::Greater,
+            (false, true) => Ordering::Less,
+            (_, _) => Ordering::Equal,
+        };
+        // Then we compare how many fields are defined by each directive.
+        let field_nums = || self.fields.len().cmp(&other.fields.len());
+        // Finally, we fall back to lexicographicaal ordering if the directives are
+        // equally specific. Although this is no longer semantically important,
+        // we need to define a total ordering to determine the directive's place
+        // in the BTreeMap.
+        let lexicographical = || {
+            self.target
+                .cmp(&other.target)
+                .then_with(|| self.in_span.cmp(&other.in_span))
+                .then_with(|| FieldOrdering(&self.fields).cmp(&FieldOrdering(&other.fields)))
+        };
+
+        let ordering = target_lengths
+            .then_with(has_span_name)
+            .then_with(field_nums)
+            .then_with(lexicographical);
```

Comment:
I'd honestly put everything inline here and just put the comments between the segments of the call chain, just so you don't have to worry about the priority falling out of sync.